File: lld_game_design_primer/minesweeper/Approach_1/Board.py
from enum import Enum

from  Cell import Cell


class GameState(Enum):
    INPROGRESS = 0
    WON = 1
    LOST = 2


class Board:
    def __init__(self, rows, columns, numberOfBombsToBePlanted):
        self._cells = [[]]
        self._unexposedCellsRemaining = 0

        self._rows = rows
        self._columns = columns
        bombs = [None for _ in range(numberOfBombsToBePlanted)]
        self._initializeBoard()
        self._plantBombsAtRandomPositions(numberOfBombsToBePlanted, bombs)
        self._setNumberedCells(bombs)
        self._unexposedCellsRemaining = rows * columns - numberOfBombsToBePlanted

    def _initializeBoard(self):
        self._cells = [[] for _ in range(self._rows)]
        r = 0
        while r < self._rows:
            c = 0
            while c < self._columns:
                self._cells[r][c] = Cell(r, c)
                c += 1
            r += 1

    def _plantBombsRandomly(self, numberOfBombsToBePlanted, bombs):
        self._plantBombsSequentially(numberOfBombsToBePlanted, bombs)
        self._shuffleBombs()

    def _shuffleBombs(self):
        nCells = self._rows * self._columns
        for index1 in range(0, nCells):
            index2 = index1 + random.randint(nCells - index1)
            if index1 != index2:
                
                row1 = index1 / self._columns
                column1 = int(math.fmod((index1 - row1 * self._columns), self._columns))
                cell1 = self._cells[row1][column1]
                
                row2 = index2 / self._columns
                column2 = int(math.fmod((index2 - row2 * self._columns), self._columns))
                cell2 = self._cells[row2][column2]
                self._cells[row1][column1] = cell2
                cell2.setRowAndColumn(row1, column1)
                self._cells[row2][column2] = cell1
                cell1.setRowAndColumn(row2, column2)

    def _plantBombsSequentially(self, numberOfBombs, bombs):
        for i in range(0, numberOfBombs):
            
            r = i / self._columns
            c = int(math.fmod((i - r * self._columns), self._columns))
            self._cells[r][c].setBomb()
            bombs[i] = self._cells[r][c]

    def _setNumberedCells(self, bombs):
        directions = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]
        for bomb in bombs:
            row = bomb.getRow()
            col = bomb.getColumn()
            for direction in directions:
                r = row + direction[0]
                c = col + direction[1]
                if 0 <= r < self._rows and 0 <= c < self._columns:
                    self._cells[r][c].setNumber(self._cells[r][c].getNumber() + 1)

    def exposeCell(self, row, col):
        cell = self._cells[row][col]
        if not cell.isExposed():
            cell.expose()
            if cell.isBlank():
                self._expandBlank()
            self._unexposedCellsRemaining -= 1
            if self._unexposedCellsRemaining == 0:
                return GameState.WON
            if cell.isBomb():
                return GameState.LOST
        return GameState.INPROGRESS

    def _expandBlank(self, cell):
        directions = [[-1, -1], [-1, 0], [-1, 1], [0, -1], [0, 1], [1, -1], [1, 0], [1, 1]]
        toExplore = []
        toExplore.add(cell)
        while not len(toExplore) == 0:
            current = toExplore.remove(0)
            for direction in directions:
                r = current.getRow() + direction[0]
                c = current.getColumn() + direction[1]
                if 0 <= r < self._rows and 0 <= c < self._columns:
                    neighbor = self._cells[r][c]
                    neighbor.expose()
                    if neighbor.isBlank():
                        toExplore.add(neighbor)

    def printBoard(self):

        pass

    

File: lld_game_design_primer/minesweeper/Approach_1/Board.java
package lld_game_design_primer.minesweeper.Approach_1;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Random;

public class Board {
    public enum GameState { INPROGRESS, WON, LOST}
    private int rows;
    private int columns;
    private Cell[][] cells;
    private int unexposedCellsRemaining; // determines if a player wins


    public Board(int rows, int columns, int numberOfBombsToBePlanted) {
        this.rows = rows;
        this.columns = columns;

        Cell[] bombs = new Cell[numberOfBombsToBePlanted];

        initializeBoard();
        plantBombsAtRandomPositions(numberOfBombsToBePlanted, bombs);
        setNumberedCells(bombs);

        unexposedCellsRemaining = rows * columns - numberOfBombsToBePlanted;
    }

    private void initializeBoard() {
        cells = new Cell[rows][columns];
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < columns; c++) {
                cells[r][c] = new Cell(r, c);
            }
        }
    }

   private void plantBombsAtRandomPositions(int numberOfBombsToBePlanted, Cell[] bombs) {
        plantBombsSequentially(numberOfBombsToBePlanted, bombs);
        shuffleBombs();
    }
    private void plantBombsSequentially(int numberOfBombs, Cell[] bombs) {
        for (int i = 0; i < numberOfBombs; i++) {
            int r = i / columns;
            int c = (i - r * columns) % columns;
            cells[r][c].setBomb();
            bombs[i] = cells[r][c];
        }
    }

    private void shuffleBombs() {
        int nCells = rows * columns;
        Random random = new Random();

        for (int index1 = 0; index1 < nCells; index1++) {

            int index2 = index1 + random.nextInt(nCells - index1);

            if (index1 != index2) {

				// compute row and column index from index1 value
                int row1 = index1 / columns;
                int column1 = (index1 - row1 * columns) % columns;

                Cell cell1 = cells[row1][column1];

				// compute row and column index corresponsing to index2
                int row2 = index2 / columns;
                int column2 = (index2 - row2 * columns) % columns;

                Cell cell2 = cells[row2][column2];

				// swap
                cells[row1][column1] = cell2;
                cell2.setRowAndColumn(row1, column1);
                cells[row2][column2] = cell1;
                cell1.setRowAndColumn(row2, column2);
            }
        }
    }

    // makes a move as per player's instruction
    // and returns the state of the game: game lost, or, game still in progress, or, player won
    public GameState exposeCell(int row, int col) {
        Cell cell = cells[row][col];
        if (!cell.isExposed()) {
            cell.expose();
            if (cell.isBlank()) {
                expandBlank(cell);
            }
            unexposedCellsRemaining--;
            if (unexposedCellsRemaining == 0) {
                return GameState.WON;
            }
            if (cell.isBomb()) {
                return GameState.LOST;
            }
        }
        return GameState.INPROGRESS;
    }

   public void expandBlank(Cell cell) {
    int[][] directions = {
            {-1, -1}, {-1, 0}, {-1, 1},
            { 0, -1},          { 0, 1},
            { 1, -1}, { 1, 0}, { 1, 1}
    };

    Queue<Cell> toExplore = new LinkedList<>();
    toExplore.add(cell);

    while (!toExplore.isEmpty()) {
        Cell current = toExplore.remove();

        for (int[] direction : directions) {
            int r = current.getRow() + direction[0];
            int c = current.getColumn() + direction[1];

            if (r >= 0 && r < rows && c >= 0 && c < columns) {
                Cell neighbor = cells[r][c];

                neighbor.expose(); // we do not check if this is a bomb cell since a blank cell can never be surrounded by a bomb cell

                // we expose both blank and numbered cell but expand on only blank cells
                if (neighbor.isBlank()) {
                    toExplore.add(neighbor);
                }
            }
        }
    }
}

    public void printBoard() {
        // print the board
    }

    private void setNumberedCells(Cell[] bombs) {
        // Offsets of 8 surrounding cells
        int[][] directions = { {-1, -1}, {-1, 0}, {-1, 1},{ 0, -1},{ 0, 1},{ 1, -1}, { 1, 0}, { 1, 1} };
        for (Cell bomb : bombs) {
            int row = bomb.getRow();
            int col = bomb.getColumn();
            for (int[] direction : directions) {
                int r = row + direction[0];
                int c = col + direction[1];
                if (r >= 0 && r < rows && c >= 0 && c < columns) {
                    cells[r][c].setNumber(cells[r][c].getNumber() + 1);
                }
            }
        }
    
    }
}
